src/sentry/auth/helper.py

```diff
@@ -90,20 +90,21 @@ def _using_okta_migration_workaround(
     return has_flag and has_provider
 
 
-Identity = Mapping[str, Any]
-
-
-@dataclass(eq=True, frozen=True)
+@dataclass
 class AuthIdentityHandler:
 
     auth_provider: Optional[AuthProvider]
     provider: Provider
     organization: Organization
     request: HttpRequest
+    identity: Mapping[str, Any]
 
-    @property
-    def user(self) -> Any:
-        return self.request.user
+    def __post_init__(self) -> None:
+        self.user: User = (
+            self.request.user
+            if self.request.user.is_authenticated
+            else (self._get_user() or self.request.user)
```

Comment:
maybe worth adding a comment, had to read it a few times before it made sense 😅  -- something like 
# prefer a logged in user, try to look up a user by email, or the anonymous user

Reply:
👍 I originally had this broken out into separate if-else statements but wanted to combine it into one expression so that the type hint could be on a single assignment. But it occurred to me that, now that `_get_user` is called from only a single place, we could fold the `is_authenticated` check into it: https://github.com/getsentry/sentry/pull/30463/commits/bfa517f661684d72c9506e4eb68d18549eb30268